### ISSUE TITLE
moor_ffi isn't working on devices with armeabi-v7a on release mode

### ISSUE THREAD

**Issue by scostagabriel** [simolus3/moor#446]
__Edit by @simolus3:__ I've opened https://github.com/flutter/flutter/issues/54948 as it seems like this is a problem in Flutter or Dart. If you're affected by this, please consider upvoting that issue.

The underlying issue has since been fixed, but the fix hasn't landed on the `stable` channel yet. Please try to switch to another channel (and run `flutter clean`).

-------

I was working normally with moor_ffi until I tried to launch a release mode of my app. When I launched some test users were reporting that the database wasn't accepting new entries. I've tried everything until I switched the version from moor_ffi 0.4.0 to moor_flutter 2.1.1. 

My flutter version is Flutter version 1.15.23-pre.11. But I was on stable and was facing the same issue.

Sadly I cannot log the error because I don't have any devices with this architecture (armeabi-v7a) and on emulators the problem wasn't showing.

But I think that these are the steps to reproduce:

1. Build an app in release mode splitting the apk by abi with `flutter build apk --target-platform android-arm,android-arm64,android-x64 --split-per-abi`

2. Install the app-armeabi-v7a-release.apk in a phone with the armeabi-v7a architecture.

3. Try to insert some data on moor database.

Sorry for the lack of details about the error. I can explain more about how I was facing the issue if necessary, but I couldn't reproduce here with my devices and I don't have an easy access to the testers.

**Comment by simolus3** [comment]
Thanks for the report. I've been getting similar reports where some flutter plugins prevented using `moor_ffi` on some abis (https://github.com/simolus3/moor/issues/420#issuecomment-593834190).

Sadly I don't have a device with that architecture, and we do compile `libsqlite3.so` for `armeabi-v7a`. I'll keep this open, but there is not much I can do here since I can't really reproduce it.

If anyone stumbles across this and wants to help: It would be great if you could try to follow the [`dart:ffi` guide from the Flutter docs](https://flutter.dev/docs/development/platform-integration/c-interop) together with the packages that appear to break `moor_ffi`. If that also fails on 32 bit arm, we can safely report an issue on Flutter or the package. If it doesn't, I can take that as a reference and check what `moor_ffi` might be doing differently.

**Comment by davidmartos96** [comment]
@simolus3 I have an old phone which is armeabi-v7a and trying out the integration test flutter_db project with FFI in it, fails in all tests with this error:
```
SqliteException: bad parameter or other API misuse, unable to open database file
package:moor_ffi/src/impl/database.dart 50                  new Database.open
package:moor_ffi/src/impl/database.dart 29                  new Database.openFile
package:moor_ffi/src/vm_database.dart 39                    _VmDelegate.open
dart:async-patch/async_patch.dart 47                        _AsyncAwaitCompleter.start
package:moor_ffi/src/vm_database.dart 37                    _VmDelegate.open
package:moor/src/runtime/executor/helpers/engines.dart 246  DelegatedDatabase.ensureOpen.<fn>
===== asynchronous gap ===========================
dart:async/zone.dart 1064                                   _CustomZone.registerBinaryCallback
dart:async-patch/async_patch.dart 84                        _asyncErrorWrapperHelper
package:test_api/src/backend/invoker.dart 0                 Invoker.waitForOutstandingCallbacks.<fn>
dart:async/zone.dart 1126                                   _rootRun
dart:async/zone.dart 1023                                   _CustomZone.run
dart:async/zone.dart 1518
```

It works fine in my other phone, which is 64 bits. Anything I can do to help out?

**Comment by simolus3** [comment]
Thanks a lot, @davidmartos96! Does that happen in debug mode as well or just release? I'll see if I can reproduce this in an emulated arm-eabi Debian.

@scostagabriel Did you get a stack trace or anything like that from testers? It would be interesting to know if that was a similar API misuse problem or something like "libsqlite3.so not found".

**Comment by davidmartos96** [comment]
@simolus3 It does occur with a debug build on the stable channel. `flutter run lib/ffi.dart`

And these are the contents of the apk. It does contain `libsqlite3.so`, so it looks alright.
![image](https://user-images.githubusercontent.com/22084723/76885795-17a4cb80-6880-11ea-81c3-d450ca1b0fd2.png)

**Comment by scostagabriel** [comment]
@simolus3 I found a stacktrace on crashlytics that probably is related to this issue, but it's not the same that @davidmartos96 posted. 

`Non-fatal Exception: java.lang.Exception: SqliteException: CHECK constraint failed: moods, constraint failed
       at PreparedStatement.execute(prepared_statement.dart:91)
       at _VmDelegate._runWithArgs(vm_database.dart:64)
       at _VmDelegate.runInsert(vm_database.dart:76)
       at _ExecutorWithQueryDelegate.runInsert.<fn>(engines.dart:69)
       at BasicLock.synchronized(basic_lock.dart:32)
       at _ExecutorWithQueryDelegate._synchronized(engines.dart:22)
       at _ExecutorWithQueryDelegate.runInsert(engines.dart:67)
       at LazyDatabase.runInsert(lazy_database.dart:67)
       at InsertStatement.insert.<fn>(insert.dart:44)
       at QueryExecutor.doWhenOpened.<fn>(executor.dart:29)
       at AsyncAction._runUnary(async_action.dart:57)
       at BasicLock.synchronized(basic_lock.dart)
       at DelegatedDatabase.ensureOpen(engines.dart:232)
       at LazyDatabase.ensureOpen.<fn>(lazy_database.dart:50)
       at AsyncAction._runUnary(async_action.dart:57)
       at AsyncAction._run(async_action.dart:44)`

This is the code from the mood table (altough I think that it's not related to this table in specific becuse it fails inserting data on other tables too).

`class Moods extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get name => text().withLength(min: 1, max: 16)();
  IntColumn get icon => integer()();
  IntColumn get position => integer()();
  BoolColumn get active => boolean().withDefault(const Constant(true))();
}`

**Comment by simolus3** [comment]
Interesting. Moor will generate a `CHECK` constraint to ensure that a `boolean` column (which is represented as an `INT`) is either 0 or 1. It might be that there is some issue with `dart:ffi` and passing integers to C functions on that platform. That would also explain the misuse exception, which indicates an invalid flag passed to `sqlite3_open_v2`. It's weird that it apparently got past the opening step for you though. 

I'll try to setup a small repro that I'd expect to fail on armv7a to investigate.

**Comment by simolus3** [comment]
I just tried to run the integration tests (`lib/ffi.dart` in `extras/integration_tests/flutter_db`) on my old Galaxy S3 (GT I9300), which has an armv7-a cpu. The integration tests went through on both release and debug builds, and both on `master` and `stable`.

@davidmartos96 Which device and Android version did you use? I'm using LineageOS 14.1, which corresponds to Android 7. That might affect the results, I can try to downgrade.

**Comment by davidmartos96** [comment]
@simolus3  I have tried on both master and stable and it fails with the same error in both `SqliteException: bad parameter or other API misuse, unable to open database file`.
The device is pretty much the same as yours, it's a Galaxy S3 Neo (GT-l9301), but it doesn't have a custom ROM, it's the stock Android version 4.4.2.

**Comment by pY4x3g** [comment]
I think I am facing a similar issue. The database is working fine on two Huawei phones, but on a Moto G5 Plus (Android 8.1), the read integers from the database are completely random.
There are no errors just the integers I read from the database are completely off the values from what I wrote.

**Comment by simolus3** [comment]
gsmarena says that the Moto G5 Plus has a 64-bit ARMv8-A processor. But it's so weird that this appears to happen in different stages: On some devices we can't even `dlopen("libsqlite3.so")`, then on others we can't open the database and now it looks like we _can_ open the database, but we're inserting bogus values.

I still couldn't find a device to reproduce this, but I have a Moto G6 to try.

**Comment by britannio** [comment]
I have an issue I'm seeing through the Play Store's Pre-launch report tool.

Out of the 11 devices tested they all use **armeabi-v7a** but 3 of the devices seem to be unable to read/write from or to the database. Here are the details and logs all provided by the Play Store.

- LGE K3 2017
  - Native platform: armeabi-v7a
  - CPU Make: Qualcomm
  - CPU Model: MSM8909
  - Android version: 6.0
  - RAM: 1024 MB
  - [logcat.txt](https://storage.googleapis.com/play-mtaas-results-prod/app.cheon.app/119/lv0-23-ru_RU-portrait/logcat?GoogleAccessId=498997737111-pmn4sf5896c460sheh5fv3k1j1cq46t5@developer.gserviceaccount.com&Expires=1586612030&Signature=KPEMJ4zuuL3QBrbx3E5NL4DznHb62Ov3zjahrX1Jzfu2RWIzNEyp1CJKUzI1pzHgJaPSo1QyTvZJ7S3cVxhT/k8O0LiorLkCJ6z%2BcR96qM2/TacrASYW3rMkipUGHwkP2CoVd0EzAhsO1hSr3ncxL9N%2Bmg4mG4lFMMNeEqZFA5CR7sE1xRhXW6c8hpCqr74f4%2Bj/hzoIc1PnCEmosCTMflF/HwzD2yQaZvQzdWcii5M%2B4wUjWioafNNIxL3zNS4ix2j6YEVopA3QEUbUfDvo7C0kZxPe1uxiuoxspLqyXOZvfM0PW2PrNj01mZeBRfSU5WblSU1s46OC%2BEIEt9Y7JQ%3D%3D&response-content-disposition=attachment;+filename%3Dlogcat.txt)
- Huawei P8 Lite
  - Native platform: armeabi-v7a
  - CPU Make: HiSilicon
  - CPU Model: Hi6220
  - Android version: 5.0
  - RAM: 2048 MB
  - [logcat.txt](https://storage.googleapis.com/play-mtaas-results-prod/app.cheon.app/119/hwALE-H-21-en_GB-portrait/logcat?GoogleAccessId=498997737111-pmn4sf5896c460sheh5fv3k1j1cq46t5@developer.gserviceaccount.com&Expires=1586612279&Signature=gjdtTQyYGNHRlS09dWNC6q0iPB2HCkgCvI1XrY6/gfzDJvsnCdeAde1vd7wxHM1J8zSoOELcviUiVwPtWOHjXXC04UlTgYw%2BnxO4PAkb4jRJUJE/zxd3jgvvTB1O5/Xzj4vdTFNycF/jkPIrs0PmbPn6DW7ZrEmAycWLRti4L5ZwqrYN2dlXg9YOAfJZJvWH1PZACxPAOa0Y9rlqcuaj/Wx/0Vi1gVNWoL7xqvYpCkwuDBm0rmDdFB0eyGjYLPLmpQDCmsTVf6NE1r1wS19T8Y02pRfG4tVHR0w7H4oL7OCAly0C/QR/afV9c7Nd1jm4Sx1c9gWcWscJHx4T7y6TWA%3D%3D&response-content-disposition=attachment;+filename%3Dlogcat.txt)
- Motorola Moto G4 Play
  - Native platform: armeabi-v7a
  - CPU Make: Qualcomm
  - CPU Model: MSM8916
  - Android version: 6.0
  - RAM: 2048 MB
  - [logcat.txt](https://storage.googleapis.com/play-mtaas-results-prod/app.cheon.app/119/harpia-23-en_US-portrait/logcat?GoogleAccessId=498997737111-pmn4sf5896c460sheh5fv3k1j1cq46t5@developer.gserviceaccount.com&Expires=1586612605&Signature=LXmE4YVhByrjtaMjXouwlAzm2iHony054oc3yzcu8Ipe%2B1XX4nIFDvRMwDWTXGqu4OfNWihfIeMaHt6CkcfcYB14vT55WJQfTRrEpJ0WkJhLEN5wyvZC70uk8%2Bm/w6dKs723vHyoi/CZWdWDG7wecrIxHSKyx2hFFBf3y7CPQXYaZTAJV%2BSMnsOgea0OuLPWDPbAxUSdbeQsE5SrBEjQtwS6lO20R0s7JOeelrnrv3McsQzIQPF3e7mVgSmKbOQLw/%2Bwd12DKPWHOKHZycyTj9mvGk5f7hEdqrmV9YS%2BpV5yf0zEgcw/yiZWJum4XbNqYHvC7gD7M6msR0NSh3X5ww%3D%3D&response-content-disposition=attachment;+filename%3Dlogcat.txt)

**Comment by simolus3** [comment]
Thanks to everyone providing additional information. I still couldn't reproduce this unfortunately, but I set up a smaller repro [here](https://github.com/simolus3/arm-ffi-repro). If anyone has access to a device where `moor_ffi` doesn't work, trying to run that repro would help a lot. It will call some C functions with a similar signature to `sqlite3_open_v2` and `sqlite3_bind_int64`. If it prints an error message, it's likely a problem with `dart:ffi` that we could then report.

If you don't need `fts5`, `json1` or isolates support, a workaround would be to use `moor_flutter` instead of `moor_ffi`.

**Comment by pY4x3g** [comment]
Here is the error:

04-15 20:16:50.450  4650  4671 E flutter : [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: test_2(1) = 1. Expected 0
04-15 20:16:50.450  4650  4671 E flutter : #0      runCheck (package:arm_ffi_repro/ffi_repro.dart:41)
04-15 20:16:50.450  4650  4671 E flutter : #1      main (package:arm_ffi_repro/main.dart:7)
04-15 20:16:50.450  4650  4671 E flutter : #2      _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:241)
04-15 20:16:50.450  4650  4671 E flutter : #3      _rootRun (dart:async/zone.dart:1184)
04-15 20:16:50.450  4650  4671 E flutter : #4      _CustomZone.run (dart:async/zone.dart:1077)
04-15 20:16:50.450  4650  4671 E flutter : #5      _runZoned (dart:async/zone.dart:1619)
04-15 20:16:50.450  4650  4671 E flutter : #6      runZonedGuarded (dart:async/zone.dart:1608)
04-15 20:16:50.450  4650  4671 E flutter : #7      _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:233)
04-15 20:16:50.450  4650  4671 E flutter : #8      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:301)
04-15 20:16:50.450  4650  4671 E flutter : #9      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168)

**Comment by simolus3** [comment]
Thank you @pY4x3g! This was on a Moto G5 Plus, running Android 8.1? Can you show the output of `flutter doctor` so that I can create an issue on Flutter for this?

**Comment by pY4x3g** [comment]
Yes it was on the Moto G5 Plus, running Android 8.1, ```flutter doctor``` gives

```Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel beta, v1.17.0, on Microsoft Windows [Version 10.0.18363.720], locale de-DE)

[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[√] Android Studio (version 3.6)
[!] IntelliJ IDEA Community Edition (version 2019.3)
    X Flutter plugin not installed; this adds Flutter specific functionality.
[!] VS Code, 64-bit edition (version 1.43.0)
    X Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
[√] Connected device (1 available)

! Doctor found issues in 2 categories.```

**Comment by fnicastri** [comment]
@simolus3 
Hi, probably the issue is related to:  flutter/flutter#55827 
reverting `settings.gradle` to an older version fixed the problem.
With the new version of `settings.gradle` `libsqlite3` is not build at all.

To reproduce the issue just create a new project with Flutter master and add moor_ffi.

Thanks
Frank

**Comment by andreystavitsky** [comment]
Have same issue on Huawei Honor 8S (MediaTek Helio A22 MT6761) using moor_ffi
query
`(6371.0 * acos(cos(:location_lat * 3.1415926 / 180) * cos(db_places.latitude * 3.1415926 / 180) * cos(db_places.longitude * 3.1415926 / 180 - :location_lng * 3.1415926 / 180) + sin(:location_lat * 3.1415926 / 180) * sin(db_places.latitude * 3.1415926 / 180))) AS distance`
show correct distance in debug mode, and completely random in release mode

Xiaomi Mi 9t Pro with snap 855 works with no issues.

**Comment by fnicastri** [comment]
@simolus3  your demo project works on my machine 'cause it uses the old gradle scripts.
Check the issue flutter/flutter#55827

**Comment by simolus3** [comment]
@fnicastri Do you mean [this demo](https://github.com/simolus3/arm-ffi-repro)? I think what you're reporting might be an unrelated issue since we can open the library and it's included in the apk. The problems occur when we call functions included in that library.

**Comment by fnicastri** [comment]
> @fnicastri Do you mean [this demo](https://github.com/simolus3/arm-ffi-repro)? I think what you're reporting might be an unrelated issue since we can open the library and it's included in the apk. The problems occur when we call functions included in that library.

I had no problems with that demo on a 32bit arm.
But compiling it with  the new gradle scripts gave me the same error mentioned in https://github.com/simolus3/moor/issues/446#issuecomment-604330350.
It prevents gradle to include the library in the apk.

Changing the scripts resolved the problem in the project where I was (not anymore) using moor_ffi.

**Comment by andreystavitsky** [comment]
Huawei Honor 8S (MediaTek Helio A22 MT6761), 64 bit Cortex-A53

```
E/flutter (22466): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: test_2(1) = 1. Expected 0
E/flutter (22466): #0      runCheck (package:arm_ffi_repro/ffi_repro.dart:41)
E/flutter (22466): #1      main (package:arm_ffi_repro/main.dart:7)
E/flutter (22466): #2      _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:241)
E/flutter (22466): #3      _rootRun (dart:async/zone.dart:1184)
E/flutter (22466): #4      _CustomZone.run (dart:async/zone.dart:1077)
E/flutter (22466): #5      _runZoned (dart:async/zone.dart:1619)
E/flutter (22466): #6      runZonedGuarded (dart:async/zone.dart:1608)
E/flutter (22466): #7      _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:233)
E/flutter (22466): #8      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:301)
E/flutter (22466): #9      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168)
```

**Comment by ruthvik-github** [comment]
<img width="964" alt="Screenshot 2020-06-21 at 11 56 23 AM" src="https://user-images.githubusercontent.com/64551464/85218261-4c13b080-b3b6-11ea-9714-8621013b4512.png">

This is the settings gradle file in my project and this is the error
<img width="757" alt="Screenshot 2020-06-21 at 11 57 49 AM" src="https://user-images.githubusercontent.com/64551464/85218280-706f8d00-b3b6-11ea-9ae6-d876fa927422.png">

flutter doctor

<img width="1180" alt="Screenshot 2020-06-21 at 12 00 05 PM" src="https://user-images.githubusercontent.com/64551464/85218311-c2181780-b3b6-11ea-83a1-862a68d7d15b.png">

**Comment by ruthvik-github** [comment]
> <img alt="Screenshot 2020-06-21 at 11 56 23 AM" width="964" src="https://user-images.githubusercontent.com/64551464/85218261-4c13b080-b3b6-11ea-9714-8621013b4512.png">
> 
> This is the settings gradle file in my project and this is the error
> <img alt="Screenshot 2020-06-21 at 11 57 49 AM" width="757" src="https://user-images.githubusercontent.com/64551464/85218280-706f8d00-b3b6-11ea-9ae6-d876fa927422.png">
> 
> flutter doctor
> 
> <img alt="Screenshot 2020-06-21 at 12 00 05 PM" width="1180" src="https://user-images.githubusercontent.com/64551464/85218311-c2181780-b3b6-11ea-83a1-862a68d7d15b.png">

Setting.gradle file that is generated in new flutter version is different from the one that is previously used. 

When setting gradle file is changed to old version see below. It is working and shows no error

<img width="918" alt="Screenshot 2020-06-21 at 1 31 10 PM" src="https://user-images.githubusercontent.com/64551464/85219682-7a4bbd00-b3c3-11ea-86b0-165e909ee811.png">

**Comment by loomp** [comment]
I believe some of my users are affected by this issue as well. Is there a workaround yet, fix is on flutter master branch?

**Comment by fnicastri** [comment]
> I believe some of my users are affected by this issue as well. Is there a workaround yet, fix is on flutter master branch?

Check my reply on the issue flutter/flutter#55827

To be sure where the problem is you should inspect the APKs to check if the sqlite.so is present in the bundle or not.

If it is not present you can try this:

**Change `settings.gradle` from:**

```
include ':app'

def flutterProjectRoot = rootProject.projectDir.parentFile.toPath()

def plugins = new Properties()
def pluginsFile = new File(flutterProjectRoot.toFile(), '.flutter-plugins')
if (pluginsFile.exists()) {
    pluginsFile.withReader('UTF-8') { reader -> plugins.load(reader) }
}

plugins.each { name, path ->
    def pluginDirectory = flutterProjectRoot.resolve(path).resolve('android').toFile()
    include ":$name"
    project(":$name").projectDir = pluginDirectory
}
```

**to:**

```
include ':app'

def localPropertiesFile = new File(rootProject.projectDir, "local.properties")
def properties = new Properties()

assert localPropertiesFile.exists()
localPropertiesFile.withReader("UTF-8") { reader -> properties.load(reader) }

def flutterSdkPath = properties.getProperty("flutter.sdk")
assert flutterSdkPath != null, "flutter.sdk not set in local.properties"
apply from: "$flutterSdkPath/packages/flutter_tools/gradle/app_plugin_loader.gradle"
```

**Comment by loomp** [comment]
Thank you @fnicastri, I will try it.
However, even with original `settings.gradle` I still have all libraries included in the lib directory (inside aab).

![image](https://user-images.githubusercontent.com/59165289/86237029-53e40980-bb9b-11ea-805e-fba07438dd2c.png)

All the devices which have issues of working with the database are of `armeabi-v7a` architecture.
The weird thing is I do not receive any errors in logs reg. missing library, but the queries are simply empty for these devices.

Anyway, I just returned from the shop with Alcatel 1B, which is based on `armeabi-v7a` architecture to finally debug it.

**Comment by fnicastri** [comment]
@loomp 
When I had this problem it was related to the absence of the library in the bundle and I resolved with that mod.
It was working perfectly.
After that I changed library because I prefer to declare my models in a more idiomatic way and then remap them on the DB, something not possible with Moor.

good luck

**Comment by loomp** [comment]
I have fixed it by changing flutter to master branch. Indeed, there was a
big in Dart as simolus3 has mentioned.

Unfortunately, fix is not released to stable branch just yet.

**Comment by andreystavitsky** [comment]
Fixed in 1.20.0
Finally.
